runtime/lua/vim/lsp.lua

```diff
@@ -211,6 +211,193 @@ local function reuse_client_default(client, config)
+lsp.config = setmetatable({ _configs = {} }, {
+  --- @param self vim.lsp.config
+  --- @param name string
+  --- @return vim.lsp.Config
+  __index = function(self, name)
+    validate('name', name, 'string')
+    self._configs[name] = self._configs[name] or {}
+    return self._configs[name]
+  end,
+
+  --- @param self vim.lsp.config
+  --- @param name string
+  --- @param cfg vim.lsp.Config
+  __newindex = function(self, name, cfg)
+    self._configs[name] = vim.tbl_deep_extend('force', self._configs[name] or {}, cfg)
+  end,
+
+  --- @param self vim.lsp.config
+  --- @param name string
+  --- @param cfg vim.lsp.Config
+  __call = function(self, name, cfg)
+    validate('name', name, 'string')
+    validate('cfg', cfg, 'table')
+    self[name] = cfg
+  end,
+})
+
+local client_config_list_keys = {
+  'filetypes',
+  'on_attach',
+  'on_exit',
+  'on_init',
+  'root_markers',
+}
+
+--- @param ... vim.lsp.Config
+--- @return vim.lsp.Config
+local function merge_configs(...)
+  local config = vim.tbl_deep_extend('force', ...)
+  for _, k in ipairs(client_config_list_keys) do
+    for i = 1, select('#', ...) do
+      local v = select(i, ...)
+      if v and v[k] then
+        config[k] = vim.list_extend(vim.deepcopy(vim._ensure_list(v[k])), config[k] or {})
+      end
+    end
+  end
+  return config
+end
+
+--- Enable a LSP server for filetype(s) to automatically start when opening a buffer.
+---
+--- Uses default configuration defined with `vim.lsp.config`.
+---
+--- Examples:
+---
+--- ```lua
+---   vim.lsp.enable('clangd', {
+---     cmd = { 'clangd', '--clang-tidy' },
+---     root_markers = { '.clangd', 'compile_commands.json' },
+---     filetypes = { 'c', 'cpp' },
+---   })
+---
+---   -- or if config is already defined
+---   vim.lsp.enable('clangd')
+---
+---   vim.lsp.enable('lua_ls', {
+---     cmd = { 'lua-language-server' },
+---     filetypes = { 'lua' },
+---     root_markers = { '.luarc.json', '.luacheckrc', '.stylua.toml' },
+---     settings = {
+---       Lua = {
+---         hint = {
+---           enable = true,
+---           paramName = 'Literal',
+---           setType = true,
+---         },
+---       },
+---     },
+---   })
+--- ```
+---
+--- @param name string
+--- @param config? vim.lsp.Config
+function lsp.enable(name, config)
+  assert(name and name ~= '*', 'Invalid name')
+  validate('config', config, 'table', true)
+
+  if config then
+    lsp.config[name] = config
+  end
+
+  config = merge_configs(lsp.config['*'], lsp.config[name], config)
+  config.name = name
+
+  validate('cmd', config.cmd, { 'function', 'table' })
+  validate('cmd', config.reuse_client, 'function', true)
+  -- TODO(lewis6991): validate more config fields
+
+  api.nvim_create_autocmd('FileType', {
+    pattern = config.filetypes,
+    group = api.nvim_create_augroup('nvim.lsp.enable.' .. name, {}),
+    callback = function(args)
+      local bufnr = args.buf --- @type integer
+      if vim.bo[bufnr].buftype == 'nofile' then
```

Comment:
No strong reason. This is all adapted from my config.

Reply:
On second thought - I also have cases where I prevent them from attaching because some servers don't like fugitive buffers and go haywire.

The safe choice is likely `buftype == ""`. 

Question is if a dedicated parameter for `.enable` is warranted to control that, but I'd tend to leave it out for the first version and then follow up depending on if the requirement comes up more often. One can still use `lsp.start` directly in the meantime.

Reply:
Ah, that's right. I'm pretty sure I do this for gitsigns diff buffers.

Reply:
Let me link #30957 again here as another example.